shared/src/main/scala/mlscript/TyperDatatypes.scala

```diff
@@ -31,6 +31,36 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
     def rigidify(implicit lvl: Int): SimpleType = freshenAbove(level, body, rigidify = true)
   }
   
+  class MethodType(level: Int, body: SimpleType, val prts: List[TypeName], val single: Bool)
+      extends PolymorphicType(level, body) {
+    def &(that: MethodType): MethodType = {
+      require(this.level === that.level)
+      MethodType(level, this.body & that.body, (this.prts ::: that.prts).distinct, false)
+    }
+    def +(that: MethodType): MethodType =
+      if (this.prts === that.prts) that
+      else MethodType(0, errType(noProv), (this.prts ::: that.prts).distinct)
+    override def toString: Str = s"MethodType($level, $body, $prts)"
+    def copyMT(level: Int = this.level, body: SimpleType = this.body, prts: List[TypeName] = this.prts): MethodType =
+      MethodType(level, body, prts, this.single)
+  }
+  object MethodType {
+    def apply(level: Int, body: SimpleType, prt: TypeName): MethodType = new MethodType(level, body, prt :: Nil, true)
+    def apply(level: Int, body: SimpleType, prts: List[TypeName]): MethodType = new MethodType(level, body, prts, true)
+    private def apply(level: Int, body: SimpleType, prts: List[TypeName], single: Bool): MethodType =
+      new MethodType(level, body, prts, single)
+    def unapply(mt: MethodType): S[(Int, SimpleType, List[TypeName])] = S((mt.level, mt.body, mt.prts))
+  }
+
+  class AbstractConstructor(val absMths: Set[TypeName])(body: SimpleType) extends PolymorphicType(0, body) {
```

Comment:
Not a big fan of this subtype. Its semantics relies on downstream code to _remember_ to check for it, and if they forget, it will silently do the wrong thing. It's unclear why it needs to be a subtype of `PolymorphicType` at all. For instance, does the `TyperHelpers` method below make sense for it?

```scala
  def subst(ts: PolymorphicType, map: Map[SimpleType, SimpleType]): PolymorphicType = 
    PolymorphicType(ts.level, subst(ts.body, map))
```

Probably not.

Better make this and `TypeScheme` both subclasses of a new `TypeInfo` base class, and make typing environments map variable names to `TypeInfo`s, to be checked in the `Var` typing case later on.

Reply:
Addressed in 822bcd4ac635eb10660981cdb4d39cb501d3d8cb